analyzer/src/functTest/java/org/jboss/gm/analyzer/alignment/SimpleProjectFunctionalTest.java

```diff
@@ -211,4 +218,163 @@ public void verifySnapshotHandling() throws IOException, URISyntaxException, Man
             });
         });
     }
+
+    @Test
+    public void verifyAlignmentReportJson() throws IOException, URISyntaxException {
+        final Path projectRoot = tempDir.newFolder("simple-project").toPath();
+        assertThat(projectRoot).isDirectory();
+        final TestManipulationModel alignmentModel = TestUtils.align(projectRoot.toFile(),
+                projectRoot.getFileName().toString());
+        assertThat(alignmentModel).isNotNull();
+        final Path buildRoot = projectRoot.resolve("build");
+        assertThat(buildRoot).isDirectory();
+        final Path jsonFile = buildRoot.resolve(Configuration.REPORT_JSON_OUTPUT_FILE);
+        assertThat(jsonFile).isRegularFile().isReadable();
+        final Path textFile = buildRoot.resolve(Configuration.REPORT_JSON_OUTPUT_FILE
+                .replaceFirst("\\.json$", ".txt"));
+        assertThat(textFile).doesNotExist();
+        final PME jsonReport = JSONUtils.fileToJSON(jsonFile.toFile());
+        final String jsonString = FileUtils.readFileToString(jsonFile.toFile(), StandardCharsets.UTF_8);
+        final String expectedJsonString = String.format(
+                "{%n" +
+                        "  \"executionRoot\" : {%n" +
+                        "    \"groupId\" : \"org.acme.gradle\",%n" +
+                        "    \"artifactId\" : \"root\",%n" +
+                        "    \"version\" : \"1.0.1\"%n" +
+                        "  },%n" +
+                        "  \"modules\" : [ {%n" +
+                        "    \"gav\" : {%n" +
+                        "      \"groupId\" : \"org.acme.gradle\",%n" +
+                        "      \"artifactId\" : \"root\",%n" +
+                        "      \"version\" : \"1.0.1.redhat-00002\",%n" +
+                        "      \"originalGAV\" : \"org.acme.gradle:root:1.0.1\"%n" +
+                        "    }%n" +
+                        "  }, {%n" +
```

Comment:
The coordinates `org.acme.gradle:root` appear three times: 

1. Execution root.
2. Without modules (_root_ project).
3. With modules.

Is this correct?

In the text file, it appears twice (equivalent to 2 and 3 above).

Reply:
I think I may have asked this already on a resolved comment, but I changed some things to add the missing `originalGav` field and the `version` field was referencing the original instead of the new one. I am sure that you're verifying that the output of both files looks good as part of the review already.

Reply:
I've checked the textual output i.e.
```
[AlignmentTask.java:729] In module root with GAV io.undertow:undertow-core:2.0.15.Final found a replacement version of io.undertow:undertow-core:2.0.15.Final-redhat-00001
[AlignmentTask.java:754] 
------------------- project org.acme.gradle:root
	Project version : 1.0.1 --> 1.0.1.redhat-00002

------------------- project org.acme.gradle:root
	Project version : 1.0.1 --> 1.0.1.redhat-00002

	Dependencies : com.yammer.metrics:metrics-core:2.2.0 --> com.yammer.metrics:metrics-core:2.2.0-redhat-00001
	Dependencies : org.hibernate:hibernate-core:5.3.7.Final --> org.hibernate:hibernate-core:5.3.7.Final-redhat-00001
	Dependencies : io.undertow:undertow-core:2.0.15.Final --> io.undertow:undertow-core:2.0.15.Final-redhat-00001


[AlignmentTask.java:658] Writing to file /tmp/junit5061978604818961430/simple-project/build/alignmentReport.json
```
and that looks good - by the way shouldn't you also assert this format/contents?


As for the `alignmentReport.json`, you are correct this does look strange - I am seeing
```
{
    "executionRoot" : {
        "groupId" : "org.acme.gradle",
        "artifactId" : "root",
        "version" : "1.0.1.redhat-00002",
        "originalGAV" : "org.acme.gradle:root:1.0.1"
    },
    "modules" :
    [
        {
            "gav" :
            {
                "groupId" : "org.acme.gradle",
                "artifactId" : "root",
                "version" : "1.0.1.redhat-00002",
                "originalGAV" : "org.acme.gradle:root:1.0.1"
            }
        },
        {
            "gav" :
            {
                "groupId" : "org.acme.gradle",
                "artifactId" : "root",
                "version" : "1.0.1.redhat-00002",
                "originalGAV" : "org.acme.gradle:root:1.0.1"
            },
            "dependencies" :
            {
                "com.yammer.metrics:metrics-core:2.2.0" : {
                    "groupId" : "com.yammer.metrics",
                    "artifactId" : "metrics-core",
                    "version" : "2.2.0-redhat-00001"
                },
                "io.undertow:undertow-core:2.0.15.Final" : {
                    "groupId" : "io.undertow",
                    "artifactId" : "undertow-core",
                    "version" : "2.0.15.Final-redhat-00001"
                },
                "org.hibernate:hibernate-core:5.3.7.Final" : {
                    "groupId" : "org.hibernate",
                    "artifactId" : "hibernate-core",
                    "version" : "5.3.7.Final-redhat-00001"
                }
            }
        }
    ]
}
```
The first GAV looks wrong to me? Does PME do this?

Reply:
I think the problem is in `if (!originalVersion.equals(newVersion)) {` block - you are adding an extra modulesitem in there. Further, I am not convinced on this version check - surely the execution root should **always** be added ?

Reply:
> and that looks good - by the way shouldn't you also assert this format/contents?

It does, unless I misunderstand what you're asking. I could change it to a `String` instead of `List<String>`, but `String` wasn't saying which line had a problem. Can I upgrade `assertj-core` to latest version for better error messages when comparing strings?

>  Improvements
> 
> Change `isEqualTo` error message to follow pattern:
> 
> expected: "abc"
> but was : "bcd"

Reply:
I think we are talking at cross-purposes - I meant checking the console output _as well as_ the json string

Reply:
> I think we are talking at cross-purposes - I meant checking the console output _as well as_ the json string

I can check if `systemOutRule.getLog()` contains the same contents as the text file. In the first test, though, no text file is created, as that tests the default settings, but I can still check the log.

Reply:
> Further, I am not convinced on this version check - surely the execution root should **always** be added ?

I think the `if` was only meant to be around the line that is printing it to the screen/text file. As far as the JSON, yes, it should always be there.

Reply:
I see the new tests - great! 

I still don't think the block `if (configuration.versionModificationEnabled()) {` is correct - as if versionModification for the project is disabled (for some reason) the entire report is lost. And I think you missed my comment about the incorrect json, and potentially why it was incorrect?

Reply:
> The first GAV looks wrong to me? Does PME do this? In fact, could that JSON even be read back in - as it doesn't look like it conforms?

I thought it was strange that `org.acme.gradle:root:1.0.1-redhat-00002` is shown twice (as `module` in JSON report and `project` in text file).

The first one is coming from `cache.getModel()` which I think is `project.getRootProject()` and the second is coming from processing the dependencies `cache.getDependencies()`. You can see it appears twice in the text file as well.

Reply:
> if (configuration.versionModificationEnabled()) {

Actually, there is no tests for this option currently. Getting at the configuration requires bmunit? See <https://github.com/project-ncl/gradle-manipulator/blob/36a6bef11661cd83042bb1cc63a86a38dd1b1c13/analyzer/src/functTest/java/org/jboss/gm/analyzer/alignment/SimpleExistingManipulationProjectFunctionalTest.java#L90-L105>.

I think it is my mistake in thinking that when this option is enabled no version change is expected to occur, but there are still potential overrides, etc. So, this just refers to the intra-project versions?

Reply:
| So, this just refers to the intra-project versions?
Yes.

| I thought it was strange that org.acme.gradle:root:1.0.1-redhat-00002 is shown twice
Its appearing 3 times

Reply:
If you count `executionRoot`, but I think this one is correct. It appears only twice in the text file. I made another comment on the code itself which might be the cause.

Reply:
> I still don't think the block `if (configuration.versionModificationEnabled()) {` is correct - as if versionModification for the project is disabled (for some reason) the entire report is lost. And I think you missed my comment about the incorrect json, and potentially why it was incorrect?

I have removed the `if (configuration.versionModificationEnabled())` entirely from that section meaning that the `executionRoot` is always set.

Reply:
> I think the problem is in `if (!originalVersion.equals(newVersion)) {` block - you are adding an extra modulesitem in there. Further, I am not convinced on this version check - surely the execution root should **always** be added ?

I have changed this so that now the `originalGav` field is set only if `!originalVersion.equals(newVersion)`, but everything else is created/set unconditionally.